### PR TITLE
WT-6488 Fail update restore eviction if no new updates were written

### PR DESCRIPTION
**Pull request by sulabhM**
<>>

### CODE REVIEW

**Comment by agorrod** [comment]
Should the title be against WT-6488?

**Comment by agorrod** [comment]
@tetsuo-cpp I've pushed a change here that helped in my local testing (the Python tests that were failing are passing, and pytpcc performed OK again). I haven't done extensive local testing - but hopefully, this is enough to get our Python suite passing while keeping the performance win.

**Comment by agorrod** [comment]
I guess I should have looked further - I think the change I pushed fixed up test_hs12, but test_hs06 and test_hs09 are still failing. It gives me hope that there is a path through though.

**Comment by keithbostic** [comment]
@sulabhM, @tetsuo-cpp, the interesting change is 2549586.

To be clear, I don't understand parts of the change, so this may be the wrong approach entirely.

**Comment by sulabhM** [comment]
Thanks @keithbostic , This change looks correct to me, I did not realize that we might be writing the overflow keys already and not referencing them again.

**Comment by tetsuo-cpp** [comment]
I gave `py-tpcc` a few more runs and it's looking good. Not totally stable but I think this is about as good as we've consistently seen.

Screen Shot 2020-07-08 at 11 18 06 am
Screen Shot 2020-07-08 at 11 34 48 am

File: src/reconcile/rec_row.c [hunk 1]
```diff
@@ -568,8 +568,27 @@ __rec_row_leaf_insert(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins)
...
+        if ((upd = upd_select.upd) == NULL) {
+            /*
+             * In cases where a page has grown so large we are trying to force evict it (there is
+             * content, but none of the content can be evicted), we set up fake split points, to
+             * allow the page to use update restore eviction and be split into multiple reasonably
+             * sized pages. Check if we are in this situation. The call to split with zero
+             * additional size is odd, but split takes into account saved updates in a special way
+             * for this case already.
+             */
+            if (!upd_select.upd_saved || !F_ISSET(r, WT_REC_EVICT) || !__wt_rec_need_split(r, 0))
```
Comment (keithbostic) [inline src/reconcile/rec_row.c:580]: I don't understand the reason for testing `WT_REF_EVICT` in this clause.

File: src/txn/txn.c [hunk 2]
```diff
@@ -2043,8 +2043,13 @@ __wt_txn_is_blocking(WT_SESSION_IMPL *session, bool conservative)
...
      *  * Hasn't done many updates
      *  * Is in the middle of a commit or abort
+     *
+     * This threshold that we're comparing the number of updates to is related and must be greater
+     * than the threshold we use in reconciliation's "need split" helper. If we're going to rollback
+     * a transaction, we need to have considered splitting the page in the case that its updates are
+     * on a single page.
      */
-    if (conservative && (txn->mod_count < 10 || F_ISSET(session, WT_SESSION_RESOLVING_TXN)))
+    if (conservative && (txn->mod_count < 20 || F_ISSET(session, WT_SESSION_RESOLVING_TXN)))
```
Comment (keithbostic) [inline src/txn/txn.c:2052]: If there is a relationship here, don't we need to use something other than a hard-coded constant?

File: src/reconcile/rec_row.c (hunk 1 above)
Comment (agorrod) [reply src/reconcile/rec_row.c:580]: Not any more, I think. This was my first attempt to fix teh overflow key case - I saw that it was in checkpoint, and I figured that we shouldn't be making checkpoint reconciliations split and do update-restore.

Comment (tetsuo-cpp) [reply src/reconcile/rec_row.c]: 130e2af

File: src/txn/txn.c (hunk 2 above)
Comment (tetsuo-cpp) [reply src/txn/txn.c]: a615eeb

**Comment by sulabhM** [comment]
@keithbostic , @tetsuo-cpp , @agorrod , another thing that I noticed while tracing how the code flowed in gdb when we are splitting empty pages:

Let's say a page split into 20 chunks. __rec_split_finish_process_prev() checks for the last two chunks 19th and 20th, if they can be combined into one and still be smaller than the page size, it will do that work. So last two chunks will be combined and that split won't happen. I think that is not a big issue because instead of 20 we will have 19 new pages. But if this were to be a page split into 2, it will reverse that work. Probably this is not a big deal, but as an enhancement, if we know that we are splitting pages, we can choose not to merge the last two chunks.

**Comment by agorrod** [comment]
> Let's say a page split into 20 chunks. __rec_split_finish_process_prev() checks for the last two chunks 19th and 20th, if they can be combined into one and still be smaller than the page size, it will do that work. So last two chunks will be combined and that split won't happen. I think that is not a big issue because instead of 20 we will have 19 new pages. But if this were to be a page split into 2, it will reverse that work. Probably this is not a big deal, but as an enhancement, if we know that we are splitting pages, we can choose not to merge the last two chunks.

Could you add that context to WT-6516 - I opened that up as a follow on to this ticket, and it's exactly the type of symptom we'd like to avoid by figuring out a better way of writing this check, or adding the check somewhere else in reconciliation.

**Comment by tetsuo-cpp** [comment]
@agorrod @sulabhM 
Patch build is looking good. Can you confirm that you're happy with this change?
As soon as stress testing goes green, I'm planning to get this in.

**Pull request merged**